src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/threadpool/QueueRejectionRca.java

```diff
@@ -0,0 +1,190 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This RCA reads ThreadPool_RejectionReqs from upstream metrics and maintain collectors for each
+ * thread pool queue type(currently we only support write/search queue). Each collector keeps track of
+ * the time window period(tp) where we repeatedly see rejections for the last tp duration.
+ * This RCA is marked as unhealthy if tp we find tp is above the threshold(300 seconds).
+ */
+public class QueueRejectionRca extends Rca<ResourceFlowUnit<HotNodeSummary>> {
+  private static final Logger LOG = LogManager.getLogger(QueueRejectionRca.class);
+  private static final long REJECTION_TIME_PERIOD_IN_MILLISECOND = TimeUnit.SECONDS.toMillis(300);
+  private final int rcaPeriod;
+  private final List<QueueRejectionCollector> queueRejectionCollectors;
+  private int counter;
+  protected Clock clock;
+
+  public <M extends Metric> QueueRejectionRca(final int rcaPeriod, M threadPool_RejectedReqs) {
+    super(5);
+    this.rcaPeriod = rcaPeriod;
+    counter = 0;
+    clock = Clock.systemUTC();
+    queueRejectionCollectors = Collections.unmodifiableList(Arrays.asList(
+        new QueueRejectionCollector(ThreadPoolEnum.WRITE_QUEUE, ThreadPoolType.WRITE,
+            threadPool_RejectedReqs, REJECTION_TIME_PERIOD_IN_MILLISECOND),
+        new QueueRejectionCollector(ThreadPoolEnum.SEARCH_QUEUE, ThreadPoolType.SEARCH,
+            threadPool_RejectedReqs, REJECTION_TIME_PERIOD_IN_MILLISECOND)
+    ));
+  }
+
+  @VisibleForTesting
+  public void setClock(Clock clock) {
+    this.clock = clock;
+  }
+
+  @Override
+  public ResourceFlowUnit<HotNodeSummary> operate() {
+    counter += 1;
+    long currTimestamp = clock.millis();
+    // collect rejection metrics
+    for (QueueRejectionCollector collector : queueRejectionCollectors) {
+      collector.collect(currTimestamp);
+    }
+    if (counter == rcaPeriod) {
+      counter = 0;
+      ClusterDetailsEventProcessor.NodeDetails currentNode = ClusterDetailsEventProcessor
+          .getCurrentNodeDetails();
+      HotNodeSummary nodeSummary = null;
+      for (QueueRejectionCollector collector : queueRejectionCollectors) {
+        // if we've see thread pool rejection in the last 5 mins, the thread pool is considered as contended
+        if (collector.isUnhealthy(currTimestamp)) {
+          if (nodeSummary == null) {
+            nodeSummary = new HotNodeSummary(currentNode.getId(), currentNode.getHostAddress());
+          }
+          nodeSummary.appendNestedSummary(collector.generateSummary(currTimestamp));
+        }
+      }
+      ResourceContext context;
+      if (nodeSummary == null) {
+        context = new ResourceContext(Resources.State.HEALTHY);
+      }
+      else {
+        context = new ResourceContext(Resources.State.UNHEALTHY);
+      }
+      boolean isDataNode = !currentNode.getIsMasterNode();
```

Comment:
In a dedicated master cluster, if something is not master, it maybe a standby master or a datatnode. If we are interested in data nodes only, then we should check for `getRole() == NodeRole.DATA`.